radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/network/ControlledSender.java

```diff
@@ -0,0 +1,190 @@
+	}
+
+	@Override
+	public void broadcastProposal(Proposal proposal, Set<BFTNode> nodes) {
+		MessageRank rank = messageRank(proposal);
+		for (BFTNode node : nodes) {
+			int receiver = this.network.lookup(node);
+			handleMessage(rank, new ControlledMessage(this.senderIndex, receiver, proposal));
+		}
+	}
+
+	@Override
+	public void sendNewView(NewView newView, BFTNode newViewLeader) {
+		int receiver = this.network.lookup(newViewLeader);
+		handleMessage(messageRank(newView), new ControlledMessage(this.senderIndex, receiver, newView));
+	}
+
+	@Override
+	public void sendVote(Vote vote, BFTNode leader) {
+		int receiver = this.network.lookup(leader);
+		handleMessage(messageRank(vote.getVoteData().getProposed(), 0), new ControlledMessage(this.senderIndex, receiver, vote));
+	}
+
+	@Override
+	public void epochChange(EpochChange epochChange) {
+		handleMessage(messageRank(epochChange), new ControlledMessage(this.senderIndex, this.senderIndex, epochChange));
+	}
+
+	@Override
+	public void sendCommittedVertex(Vertex vertex) {
+		// Ignore committed vertex signal
+	}
+
+	@Override
+	public void highQC(QuorumCertificate qc) {
+		// Ignore high QC signal
+	}
+
+	@Override
+	public void sendCommittedStateSync(long stateVersion, Object opaque) {
+		CommittedStateSync committedStateSync = new CommittedStateSync(stateVersion, opaque);
+		handleMessage(MessageRank.EARLIEST_POSSIBLE, new ControlledMessage(this.senderIndex, this.senderIndex, committedStateSync));
+	}
+
+	@Override
+	public void scheduleTimeout(LocalTimeout localTimeout, long milliseconds) {
+		handleMessage(messageRank(localTimeout), new ControlledMessage(this.senderIndex, this.senderIndex, localTimeout));
+	}
+
+	@Override
+	public void sendGetEpochRequest(BFTNode node, long epoch) {
+		// TODO: Implement
+		throw new UnsupportedOperationException();
+	}
+
+	@Override
+	public void sendGetEpochResponse(BFTNode node, VertexMetadata ancestor) {
+		// TODO: Implement
+		throw new UnsupportedOperationException();
+	}
+
+	private static class ControlledGetVerticesRequest implements GetVerticesRequest {
+		private final Hash id;
+		private final int count;
+		private final Object opaque;
+		private final int requestor;
+
+		private ControlledGetVerticesRequest(Hash id, int count, int requestor, Object opaque) {
+			this.id = id;
+			this.count = count;
+			this.requestor = requestor;
+			this.opaque = opaque;
+		}
+
+		@Override
+		public Hash getVertexId() {
+			return id;
+		}
+
+		@Override
+		public int getCount() {
+			return count;
+		}
+
+		@Override
+		public String toString() {
+			return String.format("%s{count=%s}", this.getClass().getSimpleName(), count);
+		}
+	}
+
+	private void handleMessage(MessageRank eav, ControlledMessage controlledMessage) {
+		this.network.handleMessage(eav, controlledMessage);
+	}
+
+	private MessageRank messageRank(EpochChange epochChange) {
```

Comment:
Seems to me that MessageRank shouldn't be part of ControlledSender as it's more of a communication adversary role. But fine for now but perhaps add a comment that this should move out of this class.

Reply:
These are all defaults, and message mutators can change them as they see fit, so leaving this as-is.

Reply:
Just to follow up on this, message ranking is required for timeouts to work properly.  The processing for timeouts needs to be delayed sufficiently to allow consensus rounds to complete before the timeout message is processed.

Reply:
right makes sense, but still there is no "default" to the timeout, its simply whatever the communication adversary is. But what we have here is fine for now.